From 42eccc7ccf373652dc025c947b23abea28ab384c Mon Sep 17 00:00:00 2001
From: Puyush Gupta <p-gupta5@ti.com>
Date: Sun, 17 May 2026 17:09:37 +0530
Subject: [PATCH] [kernel]: j722s: Revert "arm64: dts: ti: k3-j722s: fix
 pinctrl settings"

This reverts commit 441aa9b97ab97eca4ccf04bb5501e162765c331d in TI
kernel.

Background:
The original upstream commit removed pinctrl-single,gpio-range property
from main_pmx0 to fix issues where GPIO requests were inadvertently
resetting pin configurations. While correct for standard Linux GPIO
usage, this breaks the J722S RTOS-centric camera architecture.

Problem:
On J722S, camera capture and SerDes (FPDLink) functionality are driven
entirely by the RTOS domain. The CSI2 Expansion Interface Reset line
(CSI2_EXP_RSTz/GPIO0_68) must be muxed to GPIO mode and driven high for
camera operation. The RTOS camera driver currently does not support GPIO
pinmux configuration and relies on Linux to set up the pin muxing.

Without pinctrl-single,gpio-range:
- GPIO requests from Linux no longer automatically configure pin muxing
- The reset pin remains in its default (non-GPIO) state
- Camera initialization fails as the deserializer stays in reset
- RTOS cannot access properly muxed GPIO pins

Thus, Restore pinctrl-single,gpio-range to main_pmx0 to re-enable
automatic GPIO muxing on gpio_request(). This is a temporary solution
until RTOS camera drivers support GPIO pinmux configuration natively.

Upstream-Status: Inappropriate [customization for edgeai]

Signed-off-by: Puyush Gupta <p-gupta5@ti.com>
---
 arch/arm64/boot/dts/ti/k3-j722s-main.dtsi | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/arch/arm64/boot/dts/ti/k3-j722s-main.dtsi b/arch/arm64/boot/dts/ti/k3-j722s-main.dtsi
index 936923b9afdd..3d967d3ec0f4 100644
--- a/arch/arm64/boot/dts/ti/k3-j722s-main.dtsi
+++ b/arch/arm64/boot/dts/ti/k3-j722s-main.dtsi
@@ -471,6 +471,21 @@ &inta_main_dmss {
 	ti,interrupt-ranges = <7 71 21>;
 };
 
+&main_pmx0 {
+	pinctrl-single,gpio-range =
+		<&main_pmx0_range 0 32 PIN_GPIO_RANGE_IOPAD>,
+		<&main_pmx0_range 33 38 PIN_GPIO_RANGE_IOPAD>,
+		<&main_pmx0_range 72 17 PIN_GPIO_RANGE_IOPAD>,
+		<&main_pmx0_range 101 25 PIN_GPIO_RANGE_IOPAD>,
+		<&main_pmx0_range 137 5 PIN_GPIO_RANGE_IOPAD>,
+		<&main_pmx0_range 143 3 PIN_GPIO_RANGE_IOPAD>,
+		<&main_pmx0_range 149 2 PIN_GPIO_RANGE_IOPAD>;
+
+	main_pmx0_range: gpio-range {
+		#pinctrl-single,gpio-range-cells = <3>;
+	};
+};
+
 &main_gpio0 {
 	gpio-ranges = <&main_pmx0 0 0 32>, <&main_pmx0 32 33 38>,
 			<&main_pmx0 70 72 17>;
-- 
2.34.1

